cpmpy/transformations/int2bool.py

```diff
@@ -0,0 +1,469 @@
+        dom_size = _dom_size(x)
+        if dom_size == 0:
+            raise EMPTY_DOMAIN_ERROR
+
+        bits = math.ceil(math.log2(dom_size))
+        if bits == 0:
+            xs = cp.cpm_array([])
+        elif bits == 1:
+            xs = cp.cpm_array([boolvar(name=f"EncBin({x.name})")])
+        else:
+            xs = boolvar(shape=bits, name=f"EncBin({x.name})")
+        super().__init__(x, xs)
+
+    def encode_domain_constraint(self):
+        """Return binary encoding domain constraint (i.e. upper bound is respected with `self._x<=self._x.ub`. The lower bound is automatically enforced by offset binary which maps `000.. = self._x.lb`)."""
+        # encode directly to avoid bounds check for this seemingly tautological constraint
+        return self.encode_comparison("<=", self._x.ub, check_bounds=False)
+
+    def _bitstring(self, d):
+        """Return offset binary representation of `d` as Booleans in order of increasing signicance (e.g. `4` return `001`)."""
+        # more efficient implementation probably not necessary
+        i = self._offset(d)
+        if (
+            i == 0
+        ):  # otherwise the bitstring will formatted to '0' (which is inconsistent with empty encodings of constants)
+            return []
+        else:
+            return (
+                True if bit == "1" else False for bit in reversed("{0:b}".format(i))
+            )
+
+    def eq(self, d):
+        """Returns a list of literals which in conjunction enforced x==d."""
+        if self._x.lb == d == self._x.ub:
+            return [BoolVal(True)]
+        elif self._x.lb <= d <= self._x.ub:
+            # x_i = bit_i for every bit in the representation
+            return [
+                x if bit else (~x)
+                for bit, x in itertools.zip_longest(self._bitstring(d), self._xs)
+            ]
+        else:  # don't use try IndexError since negative values wrap
+            return [BoolVal(False)]
+
+    def encode_comparison(self, cmp, d, check_bounds=True):
+        if cmp == "==":  # x>=d and x<d+1
+            return self.eq(d)
+        elif cmp == "!=":  # x<d or x>=d+1
+            return [cp.any(~lit for lit in self.eq(d))]
+            # return [cp.any(~x for x in self.eq(d))]
+        elif cmp in (">=", "<="):
+            # TODO lexicographic encoding might be more effective, but currently we just use the PB encoding
+            constraint, domain_constraints = _encode_linear(
+                {self._x.name: self}, [self._x], cmp, d, None, check_bounds=check_bounds
+            )
+            assert (
+                domain_constraints == []
+            ), f"{self._x} should have already been encoded, so no domain constraints should be returned"
+            return constraint
+        else:
+            raise UNKNOWN_COMPARATOR_ERROR
+
+    def encode_term(self, w=1):
+        return [(w * (2**i), b) for i, b in enumerate(self._xs)], w * self._x.lb
+
+
+def _unzip(iterable):
```

Comment:
that looks odd... and unnecessary?

Reply:
Indeed only used in one place (to go from two weights/lits lists to list of 2-tuples), so I've removed it.